### ISSUE TITLE
OC4 - events overlap default functions or stop their execution

### ISSUE THREAD

**Issue by stalker780** [opencart/opencart#9409]
Just a question to Daniel. Perhaps it is not too late to modify events functionality.

Example:
I am building my own theme.

I don't want to call default catalog/controller/common/menu.php->index() from header controller.
I understand, that I can use before event for header template and collect my own $data variable. Or after event for menu->index(). But still default getCategories will be called. While I don't need this data, I plan to use my own menu builder.

How this function can be overlapped using events? Shop may have 600 of categories. Loading them takes 1-5sec from DB.

Same with default call of currency, language, cart, search controllers called from header controller.

This is just one example of hundreds possible. As far as I understand, using events I cannot stop execution of default functionality. Even if it collects useless data, which will never be used by store and may entail a lot of server resources overhead.

While with OCMOD/VQMOD this could be done with couple lines of code.
Now we need to create dozens of hardly controlled and debuggable events for all child controllers and models with hundreds lines of code.

**Comment by exelaguilar** [comment]
The only way in the current OC4 that I know you can overlap or disable an original function from being executed is to re-route it.

For example, in your theme.php file, you'd make two classes, a redirect class:

```
	public function categoryRedirect(&$route, &$data) {

		$route = 'extension/crafted/module/crafted|newCategory';

	}
```

And then the function you want to execute, or leave it empty with a return; if you don't want it to execute anything.

```
	public function newCategory() {

		// your new functionality here that will overwrite the original function

	}
```

And in your instance, you'd do an event trigger on:

catalog/controller/common/menu/before


Now it would be nice if you return anything in categoryRedirect, it overwrote the original function which looks like its how it used to or was suppose to work if we are referencing the Event System documentation (https://github.com/opencart/opencart/wiki/Events-System) but was removed or?

_**Return**
If you return the data it will stop any other actions that are_

**Comment by stalker780** [comment]
@exelaguilar 
Yes. Reroute works. I also came to this.

I created own router inside theme extension and attached it as event to all OC events catalog/view|controller|model/*/before|after.
And create dummy controllers for modules I want to disable or replace.

This looks to be flexible enough for my needs. Though it's getting even harder to debug because of lots route redirects.

I think this issue can be closed by now.
